File: internal/scanner/scanner.go
package scanner

import (
	"io/fs"
	"path/filepath"
	"strings"
	"time"
)

// FileTypeConfig 文件类型配置
type FileTypeConfig struct {
	Ext      string // 扩展名如 ".pdf"
	Type     string // 类型标识如 "pdf"
	Category string // 分类: "document" / "image" / "video"
}

// 所有支持的文件类型
var allFileTypes = []FileTypeConfig{
	// 文档
	{".pdf", "pdf", "document"},
	{".doc", "docx", "document"},
	{".docx", "docx", "document"},
	{".xls", "xlsx", "document"},
	{".xlsx", "xlsx", "document"},
	{".ppt", "pptx", "document"},
	{".pptx", "pptx", "document"},
	// 图片
	{".jpg", "image", "image"},
	{".jpeg", "image", "image"},
	{".png", "image", "image"},
	{".gif", "image", "image"},
	{".bmp", "image", "image"},
	{".webp", "image", "image"},
	{".svg", "image", "image"},
	{".ico", "image", "image"},
	{".tiff", "image", "image"},
	{".tif", "image", "image"},
	// 视频
	{".mp4", "video", "video"},
	{".avi", "video", "video"},
	{".mkv", "video", "video"},
	{".mov", "video", "video"},
	{".wmv", "video", "video"},
	{".flv", "video", "video"},
	{".webm", "video", "video"},
	{".m4v", "video", "video"},
}

// 根据启用的分类构建扩展名查找表
func buildExtMap(enabledTypes []string) map[string]string {
	extMap := make(map[string]string)
	for _, ft := range allFileTypes {
		for _, enabled := range enabledTypes {
			if ft.Type == enabled {
				extMap[ft.Ext] = ft.Type
				break
			}
		}
	}
	return extMap
}

// ScanResult 扫描结果
type ScanResult struct {
	Path     string
	Filename string
	FileType string
	FileSize int64
	ModTime  time.Time
}

// ScanFolder 递归扫描文件夹中支持的文件类型
// enabledTypes: 启用的类型列表如 ["pdf","docx","xlsx","pptx","image","video"]
func ScanFolder(folderPath string, enabledTypes []string) ([]ScanResult, error) {
	extMap := buildExtMap(enabledTypes)
	var results []ScanResult

	// 预先将folderPath转为绝对路径，避免循环内重复调用filepath.Abs
	absFolderPath, err := filepath.Abs(folderPath)
	if err != nil {
		absFolderPath = folderPath
	}

	err = filepath.WalkDir(absFolderPath, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			return nil
		}

		// 只对匹配扩展名的文件获取详细信息
		ext := strings.ToLower(filepath.Ext(d.Name()))
		fileType, ok := extMap[ext]
		if !ok {
			return nil
		}

		// 获取文件信息（只对匹配的文件调用Info()）
		info, err := d.Info()
		if err != nil {
			return nil
		}

		results = append(results, ScanResult{
			Path:     path, // WalkDir返回的路径已经是绝对路径
			Filename: d.Name(),
			FileType: fileType,
			FileSize: info.Size(),
			ModTime:  info.ModTime(),
		})
		return nil
	})

	return results, err
}

// GetSupportedTypes 返回所有支持的文件类型标识（去重）
func GetSupportedTypes() []string {
	seen := make(map[string]bool)
	var types []string
	for _, ft := range allFileTypes {
		if !seen[ft.Type] {
			seen[ft.Type] = true
			types = append(types, ft.Type)
		}
	}
	return types
}

File: src-tauri/src/scanner.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;
use walkdir::WalkDir;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ScanResult {
    pub path: String,
    pub filename: String,
    pub file_type: String,
    pub file_size: i64,
    pub mod_time: String,
}

fn get_all_extensions() -> HashMap<&'static str, Vec<&'static str>> {
    let mut m = HashMap::new();
    m.insert("pdf", vec!["pdf"]);
    m.insert("docx", vec!["docx", "doc"]);
    m.insert("xlsx", vec!["xlsx", "xls"]);
    m.insert("pptx", vec!["pptx", "ppt"]);
    m.insert("image", vec!["jpg", "jpeg", "png", "gif", "bmp", "webp", "svg"]);
    m.insert("video", vec!["mp4", "avi", "mkv", "mov", "wmv", "flv"]);
    m
}

fn build_ext_map(enabled_types: &[String]) -> HashMap<String, String> {
    let all = get_all_extensions();
    let mut ext_map = HashMap::new();
    for (type_id, exts) in &all {
        if enabled_types.contains(&type_id.to_string()) {
            for ext in exts {
                ext_map.insert(ext.to_string(), type_id.to_string());
            }
        }
    }
    ext_map
}

pub fn get_file_type(path: &str) -> String {
    let ext = Path::new(path)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_lowercase();
    let all = get_all_extensions();
    for (type_id, exts) in &all {
        if exts.contains(&ext.as_str()) {
            return type_id.to_string();
        }
    }
    "unknown".to_string()
}

pub fn scan_folder(folder_path: &str, enabled_types: &[String]) -> Vec<ScanResult> {
    let ext_map = build_ext_map(enabled_types);
    let abs_path = std::fs::canonicalize(folder_path).unwrap_or_else(|_| folder_path.into());
    let mut results = Vec::new();

    for entry in WalkDir::new(&abs_path).into_iter().filter_map(|e| e.ok()) {
        if !entry.file_type().is_file() {
            continue;
        }
        let ext = entry.path()
            .extension()
            .and_then(|e| e.to_str())
            .unwrap_or("")
            .to_lowercase();
        if let Some(file_type) = ext_map.get(&ext) {
            let metadata = entry.metadata().ok();
            let file_size = metadata.as_ref().map(|m| m.len() as i64).unwrap_or(0);
            let mod_time = metadata
                .and_then(|m| m.modified().ok())
                .map(|t| {
                    let dt: chrono::DateTime<chrono::Local> = t.into();
                    dt.format("%Y-%m-%d %H:%M:%S").to_string()
                })
                .unwrap_or_default();
            results.push(ScanResult {
                path: entry.path().to_string_lossy().to_string(),
                filename: entry.file_name().to_string_lossy().to_string(),
                file_type: file_type.clone(),
                file_size,
                mod_time,
            });
        }
    }
    results
}
